### ISSUE TITLE
Support for spread operator in JSX (without ES6)

### ISSUE THREAD

**Issue by JakeSidSmith** [eslint/eslint#6852]
Using the spread operator in JSX should work in any ecma version, but instead you must set the version to 6 (or higher).

Spread is part of the JSX spec: http://facebook.github.io/jsx/
And is supported in JSX by other libraries, such as babel (with babel-preset-react).
Acorn has an official plugin that supports this: https://github.com/RReverser/acorn-jsx
And espree's readme has an example, having support for spread in ES5: https://github.com/eslint/espree

Using `eslint --init` and answering yes to React & JSX generates config that does not allow the spread operator (as ecma version 6 is required, see this issue: https://github.com/eslint/eslint/issues/5608).

Note: babel (recommended parser) handles the spread operator perfectly in JSX with only babel-preset-react.

**What version of ESLint are you using?**

`2.8.0`

**What parser (default, Babel-ESLint, etc.) are you using?**

eslint with eslint-plugin-react

**Please show your full configuration:**

Rules do not appear to affect this issue. Only passes with ecma version 6 (or higher).

.eslintrc

``` json
{
  "parserOptions": {
    "ecmaFeatures": {
      "jsx": true,
      "experimentalObjectRestSpread": true
    }
  },
  "plugins": [
    "react"
  ]
}
```

**What did you do? Please include the actual source code causing the issue.**

Using the spread operator to pass props to a JSX element with ecma version 5 

``` javascript
<div {...this.props} />
```

**What did you expect to happen?**

This should pass the linting.

**What actually happened? Please include the actual, raw output from ESLint.**

`error  Parsing error: Unexpected token`

which points to the open brace: `{`

**Comment by platinumazure** [comment]
Thanks for the issue!

Adding "blocked" label because this is likely an upstream issue. We will probably create an issue upstream and link to that issue from here (and we may close this issue at that time).

**Comment by kaicataldo** [comment]
Not sure (didn't have much time to look), but might be relevant: https://github.com/eslint/espree/blob/master/espree.js#L316

**Comment by nzakas** [comment]
This may be an issue with the Acorn plugin that Espree uses, as we delegate to that for parsing JSX. In any event, this should be a bug on Espree. 

@kaicataldo that's for experimental object rest/spread and shouldn't be used by JSX. 

@JakeSidSmith experimental rest/spread flag isn't used for JSX parsing at all.

**Comment by nzakas** [comment]
> The config generator adds experimentalObjectRestSpread for React / JSX but this shouldn't be needed

This is the correct behavior. People using React tend to use this feature, which is why we turn it on. 

Really, the first and last issue you mention are really the same, and the third is a symptom of those. 

I don't have enough energy to investigate this, so unless someone else is willing to take a stab at it, this might take a while to resolve.

**Comment by JakeSidSmith** [comment]
I don't mind taking a look if someone can point me in the right direction. :)

**Comment by kaicataldo** [comment]
Did a little poking around to try to get you started, @JakeSidSmith.

It looks like the [acorn-jsx plugin](https://github.com/RReverser/acorn-jsx) we use in [Espree](https://github.com/eslint/espree) (our parser) supports this out of the box. Running the following
```
"use strict";

require('babel-register')({
  presets: [ 'es2015' ]
});

var acorn = require('acorn-jsx/inject')(require('acorn'));
var ast = acorn.parse(
  '<div {...this.props} />',
  {
    plugins: { jsx: true }
  }
);

console.log(JSON.stringify(ast, null, 4));
```
yielded these results:
```
{
    "type": "Program",
    "start": 0,
    "end": 23,
    "body": [
        {
            "type": "ExpressionStatement",
            "start": 0,
            "end": 23,
            "expression": {
                "type": "JSXElement",
                "start": 0,
                "end": 23,
                "openingElement": {
                    "type": "JSXOpeningElement",
                    "start": 0,
                    "end": 23,
                    "attributes": [
                        {
                            "type": "JSXSpreadAttribute",
                            "start": 5,
                            "end": 20,
                            "argument": {
                                "type": "MemberExpression",
                                "start": 9,
                                "end": 19,
                                "object": {
                                    "type": "ThisExpression",
                                    "start": 9,
                                    "end": 13
                                },
                                "property": {
                                    "type": "Identifier",
                                    "start": 14,
                                    "end": 19,
                                    "name": "props"
                                },
                                "computed": false
                            }
                        }
                    ],
                    "name": {
                        "type": "JSXIdentifier",
                        "start": 1,
                        "end": 4,
                        "name": "div"
                    },
                    "selfClosing": true
                },
                "closingElement": null,
                "children": []
            }
        }
    ],
    "sourceType": "script"
}
```

So the fix is definitely either in Espree or ESLint! I would start with Espree and see if there are any conditions in which it doesn't parse `JSXSpreadAttribute` when it should. If not, we'll probably have to take a look at ESLint itself (i.e. follow the dependency chain :D).

Thanks for being willing to look into this, and let us know if there's anything we can do to help, either here in the issue or in our [Gitter channel](https://gitter.im/eslint/eslint).

**Comment by JakeSidSmith** [comment]
Thanks for the info. Will try to take a look when I've got a moment. :)

**Comment by kaicataldo** [comment]
Sorry, sorry, just realized acorn parses ES6 by default. The above is wrong :(

```
"use strict";

require('babel-register')({
  presets: [ 'es2015' ]
});

var acorn = require('acorn-jsx/inject')(require('acorn'));
var ast = acorn.parse(
  '<div {...this.props} />',
  {
    ecmaVersion: 5,
    plugins: { jsx: true }
  }
);

console.log(JSON.stringify(ast, null, 4));
```
actually yields:
```
/Users/kaicataldo/Code/test/acorn-test/node_modules/acorn/dist/acorn.js:2223
    throw err
    ^

SyntaxError: Unexpected token (1:6)
    at Parser.pp$4.raise (/Users/kaicataldo/Code/test/acorn-test/node_modules/acorn/dist/acorn.js:2221:15)
    at Parser.pp.unexpected (/Users/kaicataldo/Code/test/acorn-test/node_modules/acorn/dist/acorn.js:603:10)
    at Parser.pp.expect (/Users/kaicataldo/Code/test/acorn-test/node_modules/acorn/dist/acorn.js:597:28)
    at Parser.pp.jsx_parseAttribute (/Users/kaicataldo/Code/test/acorn-test/node_modules/acorn-jsx/inject.js:272:12)
    at Parser.pp.jsx_parseOpeningElementAt (/Users/kaicataldo/Code/test/acorn-test/node_modules/acorn-jsx/inject.js:289:33)
    at Parser.pp.jsx_parseElementAt (/Users/kaicataldo/Code/test/acorn-test/node_modules/acorn-jsx/inject.js:310:31)
    at Parser.pp.jsx_parseElement (/Users/kaicataldo/Code/test/acorn-test/node_modules/acorn-jsx/inject.js:359:17)
    at Parser.parseExprAtom (/Users/kaicataldo/Code/test/acorn-test/node_modules/acorn-jsx/inject.js:381:23)
    at Parser.pp$3.parseExprSubscripts (/Users/kaicataldo/Code/test/acorn-test/node_modules/acorn/dist/acorn.js:1715:21)
    at Parser.pp$3.parseMaybeUnary (/Users/kaicataldo/Code/test/acorn-test/node_modules/acorn/dist/acorn.js:1692:19)
```

It does look like it's an upstream issue with either `acorn` or the `acorn-jsx` plugin...

**Comment by JakeSidSmith** [comment]
Seems like this is the culprit: https://github.com/ternjs/acorn/blob/b899480692b6fb835458b4126165b820ccfa4fea/src/tokenize.js#L215

Removing the check for ES6 runs fine with the sample @kaicataldo posted above

```javascript
var acorn = require('acorn-jsx');

var result = acorn.parse(
  '<div {...this.props} />',
  {
    ecmaVersion: 5,
    plugins: {
      jsx: true
    }
  }
);

console.log(JSON.stringify(result, null, 2))
```

Output

```json
{
  "type": "Program",
  "start": 0,
  "end": 23,
  "body": [
    {
      "type": "ExpressionStatement",
      "start": 0,
      "end": 23,
      "expression": {
        "type": "JSXElement",
        "start": 0,
        "end": 23,
        "openingElement": {
          "type": "JSXOpeningElement",
          "start": 0,
          "end": 23,
          "attributes": [
            {
              "type": "JSXSpreadAttribute",
              "start": 5,
              "end": 20,
              "argument": {
                "type": "MemberExpression",
                "start": 9,
                "end": 19,
                "object": {
                  "type": "ThisExpression",
                  "start": 9,
                  "end": 13
                },
                "property": {
                  "type": "Identifier",
                  "start": 14,
                  "end": 19,
                  "name": "props"
                },
                "computed": false
              }
            }
          ],
          "name": {
            "type": "JSXIdentifier",
            "start": 1,
            "end": 4,
            "name": "div"
          },
          "selfClosing": true
        },
        "closingElement": null,
        "children": []
      }
    }
  ]
}
```

**Comment by JakeSidSmith** [comment]
So, inside acorn-jsx I've set `insideJSXTag`... and got it working nicely, but not sure how you could handle this purely in acorn.

```javascript
tt.jsxTagStart.updateContext = function() {
    this.insideJSXTag = true;
    // etc

tt.jsxTagEnd.updateContext = function(prevType) {
    this.insideJSX = false;
    // etc
```

And then in acorn

```javascript
  pp$7.readToken_dot = function() {
    var next = this.input.charCodeAt(this.pos + 1)

    if (next >= 48 && next <= 57) return this.readNumber(true)
    var next2 = this.input.charCodeAt(this.pos + 2)
    if (
      (
        this.options.ecmaVersion >= 6 ||
        (typeof this.options.plugins.jsx !== 'undefined' && this.insideJSX)
      ) && next === 46 && next2 === 46
    ) { // 46 = dot '.'
      this.pos += 3
      return this.finishToken(tt.ellipsis)
    } else {
      ++this.pos
      return this.finishToken(tt.dot)
    }
  }
```

**Comment by kaicataldo** [comment]
Please see https://github.com/RReverser/acorn-jsx/issues/40.

>In order to change this, it would indeed require various modifications of Acorn itself (to present new ecmaVersion checks in various places of parser instead of single check in tokenizer).

**Comment by gyandeeps** [comment]
can we close this issue since upstream issue got closed?

**Comment by not-an-aardvark** [comment]
Sounds good to me.
